---
title: "BgeeCall"
author: "Alessandro Brandulas Cammarata"
date: "2025-09-05"
output: html_document
---

```{r setup, include=FALSE}
knitr::opts_chunk$set(echo = TRUE)
```

# Introduction

Genes called present in an RNA-Seq library are often defined using an abitrary cutoff of log2(TPM) = 1. BgeeCall allows to calculate a library specific cutoff. Reference intergenic sequences are the key element to define this library specific TPM cutoff. The cutoff is defined by fitting a hurdle model to the reads mapped to intergenic regions in that RNA-Seq library, using this fit as a null model, we can then estimate at which TPM value a gene is truly expressed above background noise.

## Setting up BgeeCall

### Installing BgeeCall

```{r, message=FALSE, warning=FALSE, results='hide'}
if (!require("BiocManager", quietly = TRUE))
    install.packages("BiocManager")
    BiocManager::install(version = "3.21", ask = FALSE)

```

```{r, message=FALSE, warning=FALSE, results='hide'}
BiocManager::install("BgeeCall", ask = FALSE)
library(BgeeCall)
```

```{r, message=FALSE, warning=FALSE, results='hide'}
#install.packages("R.utils")
#BiocManager::install("GenomeInfoDbData", ask = FALSE)
library(R.utils)
library(GenomeInfoDbData)
```

### Download annotation and transcriptome files for Drosophila melanogaster

```{r}
annotation_file_path =  "./data/Drosophila_melanogaster_Annotation.chr.gtf.gz"
download.file("http://ftp.ensemblgenomes.org/pub/release-62/metazoa/gtf/drosophila_melanogaster/Drosophila_melanogaster.BDGP6.54.62.gtf.gz", annotation_file_path)
gunzip(annotation_file_path)


transcriptome_file_path = "./data/Drosophila_melanogaster_Transcriptome.cdna.all.fa.gz"
download.file("http://ftp.ensemblgenomes.org/pub/release-62/metazoa/fasta/drosophila_melanogaster/dna/Drosophila_melanogaster.BDGP6.54.dna.toplevel.fa.gz", transcriptome_file_path)
gunzip(transcriptome_file_path)

# In the next release of Bgee (Bgee16), we will directly have this function to download both the gtf and fasta
##BgeeCall::retrieve_fasta_gtf_from_taxonid(taxon_id=c(7227), ensembl_metazoa_release=62, outDir="./data/")
```
### Downloading fastq files from your RNA-Seq library
```{bash}
#Download your favourite RNAseq fastq library

#Either through SRA
#prefetch SRR384929
#fasterq-dump --outdir ./data/ --split-files SRR384929

# or through their website
#https://trace.ncbi.nlm.nih.gov/Traces/?view=run_browser&acc=SRR384929&display=download

```


### Getting intergenic regions for your species

We can list all available intergenic releases by running the following function:
```{r}
list_intergenic_release()
```

### Bgee releases
All releases for which the name is a number are Bgee releases. It means all reference intergenic sequences from these releases have been generated by the Bgee team. It is possible to list all species available in one bgee release with those commands:

```{r}
bgee <- new("BgeeMetadata", intergenic_release="1.0")

species_list <- list_bgee_ref_intergenic_species(myBgeeMetadata = bgee)
species_available <- species_list$speciesName
print(species_available)
```

By looking at the full table you can also see how many libraries we have for each species in Bgee
```{r}
print(species_list)
```

### Community release
If your species of interest is not present in the last official intergenic release you can check if the BgeeCall community already generated the reference intergenic sequences. To do so use the commands
```{r}
list_community_ref_intergenic_species()
#bgee <- new("BgeeMetadata", intergenic_release="community")
##https://zenodo.org/communities/bgee_intergenic/records?q=&l=list&p=1&s=10&sort=newest
#bgee <- new("BgeeMetadata", intergenic_release="community")
```


### Generating intergenics for a new species
If both the last options didn't include your species of interest, you can use BgeeCall to generate an intergenic release for a species available in ensembl. The following functions will help you to do so:
In a future Bgee version, we aim to generate those sequences for all species in ensembl and include them in the official Bgee release.
```{r}
# if you already have selected intergenic regions for your species of interest
bgee <- new("BgeeMetadata", intergenic_release = "custom")
bgee@intergenic_release <- "custom"
user <- new("UserMetadata", species_id = "7227", reads_size=100)
user@custom_intergenic_path = "./data/7227_intergenic.fa" #example file in ./data/Caenorhabditis_elegans_intergenic1000+.fa

#Otherwise there is some function in BgeeCall to generate them from the gtf and fasta files of ensembl
# if you also need to download the gft and fasta from ensembl
#generate_intergenic_with_ensembl <- function(species_gtf = c("homo_sapiens/Homo_sapiens.GRCh38", "gallus_gallus/Gallus_gallus.bGalGal1.mat.broiler.GRCg7b"), ensembl_release = 112, ensembl_metazoa_release = 59, gtf_dir = "./data/", intergenic_dir = "./data")

# if you just need to generate the intergenic sequences
#generate_initial_intergenic_regions(gene_gtf_path="./data/Drosophila_melanogaster_Annotation.chr.gtf", genome_fasta_path="./data/Drosophila_melanogaster_Transcriptome.cdna.all.fa", N_block_size=31, N_proportion=0.05, output_gtf_path="./data/")

# We also have a function to remove intergenic regions that seem to have too many reads mapped to them that might come from unannotated genes
#generate_reference_intergenic_regions that take as input the files generated by the previous functions
```

## Generating presence/absence calls for your library

### BgeeMetadata and KallistoMetadata objects
These 2 classes are useful to tune how BgeeCall works. No need to create an object of the class BgeeMetada as only default values will be used during this practical (last official Bgee intergenic release).
An object of the KallistoMetadata class has to be created to specify to download kallisto.
kallisto will be used as our pseudo-aligner in order to map our reads to both genes and intergenic regions.
```{r}
kallisto <- new("KallistoMetadata", download_kallisto = TRUE)
```

You can retrieve the taxonid of your species by using the species_list we previously generated.
```{r}
searched_speciesId <- species_list$speciesId[species_list$speciesName == "Drosophila melanogaster"]
```


### UserMetadata object
An object of this class has to be created and value of some slots have to be modified to run BgeeCall.
```{r}
user <- new("UserMetadata", species_id = "7227", reads_size=100)
user@custom_intergenic_path = "./data/7227_intergenic.fa"
user <- setRNASeqLibPath(user, "./data/ERX2162340/")
user <- setTranscriptomeFromFile(user, "./data/Drosophila_melanogaster_Transcriptome.cdna.all.fa")
user <- setAnnotationFromFile(user, "./data/Drosophila_melanogaster_Annotation.chr.gtf")
user <- setOutputDir(user, "./output/ERX2162340/")
user <- setWorkingPath(user, "./kallisto/")

```

Now that all objects have been created it is possible to run the generation of present/absent gene expression calls with one unique line of code.
This command generates a list with a link to 5 generated files.

  calls_tsv_path : path to the file with abundance at gene level and present/absent expression calls
  cutoff_info_file_path : path to the file with metadata about TPM cutoff (TPM cutoff, protein coding genes present, etc.)
  abundance_tsv : path to the kallisto output file with abundance at transcript level
  TPM_distribution_path : path to a PDF file containing a plot of abundance distribution with a line representing the TPM cutoff.
  S4_slots_sumary : path to the file with value of most important slots of the 3 BgeeCall classes
```{r}
output_files_info <- generate_calls_workflow(abundanceMetadata = kallisto, bgeeMetadata=bgee, userMetadata = user)
output_files_info
```

## Generating calls for multiple libraries
BgeeCall uses a tsv file to generate present/absent calls for more than one library. In this file each line will be used to generate one UserMetadata object. Each column of this file correspond to the value of one slot of the object. A template of the file called inputFile.tsv is present in the input_files directory of the project.

The columns of this file are :

  species_id : The NCBI ID of the species.
  reads_size : The size of the reads of the RNA-Seq library.
  rnaseq_lib_path : Path to the directory containing fastq files of the library.
  transcriptome_path : path to the transcriptome file.
  annotation_path : path to the genome annotation file. Works with GTF of GFF3 files.
  working_path : path to the working directory where results at species level will be stored (kallisto index, transcriptome with intergenic, etc.)
  output_directory : path to directory where calls of one RNA-Seq library have to be written (e.g. /path/to/output/LIBRARY_ID/)

We put an example file with the format we need called: ./data/Drosophila_libraries.tsv

# Merging the pValues of multiple libraries to get a consensus call of expression
If we have multiple libraries that are from the same conditions (e.g. liver or testis), we can use BgeeCall to merge the p-values of expression obtained previously to have a more robust call on which genes are expressed in those conditions. That's what we do in Bgee when show the expression of a gene in a tissue, developmental stage or sex for example.
```{r}
merging_libraries(userFile = "./data/3Humanlibrary_merged.tsv", approach = "Mean", outDir = "./output/merged/", weights = TRUE)
```

